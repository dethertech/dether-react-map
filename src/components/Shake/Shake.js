import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import styled, { keyframes, css } from 'styled-components'

const shakeAnim = keyframes`
  10%, 90% {transform: translateX(-1px);}
  20%, 80% {transform: translateX(2px);}
  30%, 50%, 70% {transform: translateX(-4px);}
  40%, 60% {transform: translateX(4px);}
`

const ShakeWrapper = styled.div`
  transform: translateX(0);

  ${({ shake }) =>
    shake &&
    css`
      animation: ${shakeAnim} 0.82s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
    `};
`

/**
 * Shake component Animation
 * @extends PureComponent
 */

class Shake extends PureComponent {
  static propTypes = {
    toggle: PropTypes.number,
    children: PropTypes.oneOfType([PropTypes.array, PropTypes.element])
      .isRequired,
  }

  static defaultProps = {
    toggle: 0,
  }

  constructor(props) {
    super(props)
    this.state = {
      shake: false,
      toggle: 0,
    }
    this.interval = null
  }

  componentWillReceiveProps = ({ toggle: propToggle }) => {
    const { toggle: stateToggle } = this.state

    if (propToggle > stateToggle) {
      this.setState({ shake: true, toggle: propToggle })
      this.interval = setTimeout(
        () => this.setState({ shake: false, toggle: 0 }),
        850,
      )
    }
  }

  componentWillUnmount() {
    clearTimeout(this.interval)
  }

  render = () => {
    const { children } = this.props
    const { shake } = this.state

    return <ShakeWrapper shake={shake}>{children}</ShakeWrapper>
  }
}

export default Shake
