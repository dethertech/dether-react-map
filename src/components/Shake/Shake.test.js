/* global describe it expect */
import React from 'react';
import renderer from 'react-test-renderer';



import Shake from './index';



describe('Component Shake', () => {
  const child = <p>Child</p>;
  const component = renderer.create(<Shake toggle={5}>{child}</Shake>);
  const tree = component.toJSON();

  it('should be render without crash', () => {
    expect(tree).toMatchSnapshot();
  });
});
