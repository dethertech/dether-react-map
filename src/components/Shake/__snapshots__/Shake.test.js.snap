// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Shake should be render without crash 1`] = `
.c0 {
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
}

<div
  className="c0"
>
  <p>
    Child
  </p>
</div>
`;
